### PR TITLE
MaterialBuilder exposed to python

### PR DESCRIPTION
**Pull request by peterfpeterson**
The main goal was to expose `MaterialBuilder` to python. In the process there was a refactor of the underlying `Material` class and how it stores information. This was caused by a need to reproduce various factors used by PDF measurements.

**To test:**

This should be most verified by review and seeing that all of the tests still pass. A script that produces (in standard units) most of "gemsheet" is:

``` python
import mantid
from mantid.kernel import MaterialBuilder
import math

def printSummary(material):
    chem = material.chemicalFormula()
    print 'Sym   AMass    Num   c    b bar  b     Sigma s  Sigma a'
    totalAtoms = 0.
    for amount in chem[1]:
        totalAtoms += amount

    for (atom, amount) in zip(chem[0], chem[1]):
        neutron = atom.neutron()
        bbar = neutron['coh_scatt_length_real'] # according to gemsheet
        b = neutron['tot_scatt_length']         # according to gemsheet
        print '%3s%9.4f%6.2f%6.2f%7.4f%7.4f%8.4f%8.4f' % (atom.symbol, atom.mass,
                                                          amount,
                                                          amount/totalAtoms,
                                                          bbar,
                                                          b,
                                                          neutron['tot_scatt_xs'], # Sigma s
                                                          neutron['abs_xs']) # Sigma a

    print "(sum c*b^2)     %.4f" % mat.totalScatterLengthSqrd() # E12
    print "(sum c bbar)^2  %.4f" % (mat.cohScatterLengthReal()*mat.cohScatterLengthReal()) # E14

builder = MaterialBuilder()
mat = builder.setFormula('Ba Ti O3').build()
print '#################### Ba Ti O3'
printSummary(mat)

builder = MaterialBuilder()
mat = builder.setFormula('Al2 O3').build()
print '#################### Al2 O3'
printSummary(mat)
```

It does not generate the Faber-Ziman factors, but the information is now easily created in python.

_There is no associated issue_

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/src/Material.cpp [hunk 1]
```diff
@@ -137,15 +476,63 @@ void Material::saveNexus(::NeXus::File *file, const std::string &group) const {
...
+      file->readData("coh_scatt_xs", neutron.coh_scatt_xs);
+      file->readData("inc_scatt_xs", neutron.inc_scatt_xs);
+      file->readData("tot_scatt_xs", neutron.tot_scatt_xs);
+      file->readData("abs_scatt_xs", neutron.abs_scatt_xs);
+      file->readData("tot_scatt_length", neutron.tot_scatt_length);
+      file->readData("coh_scatt_length", neutron.coh_scatt_length);
+      file->readData("inc_scatt_length", neutron.inc_scatt_length);
+
+      FormulaUnit formulaUnit{boost::make_shared<Atom>(neutron), 1.};
+      m_chemicalFormula.push_back(formulaUnit);
```
Comment (martyngigg) [inline Framework/Kernel/src/Material.cpp]: We could eliminate the temporary using `emplace_back` if we added a constructor to Formula unit.

File: Framework/Kernel/src/Material.cpp [hunk 2]
```diff
@@ -72,8 +115,22 @@ double Material::pressure() const { return m_pressure; }
  * the given wavelength
  */
 double Material::cohScatterXSection(const double lambda) const {
-  (void)lambda;
-  return m_element.coh_scatt_xs;
+  UNUSED_ARG(lambda);
```
Comment (martyngigg) [inline Framework/Kernel/src/Material.cpp]: Could the lack of caching impact performance for the absorption correction algorithms?

File: Framework/PythonInterface/mantid/kernel/src/Exports/MaterialBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      .def("setIncoherentXSection", &MaterialBuilder::setIncoherentXSection,
+           return_self<>(), (arg("self"), arg("xsec")),
+           "Set the atomic number of the material")
+      .def("setAbsorptionXSection", &MaterialBuilder::setAbsorptionXSection,
+           return_self<>(), (arg("self"), arg("xsec")),
+           "Set the atomic number of the material")
+
+      .def("build", &MaterialBuilder::build,
+           return_value_policy<
+               return_by_value>(), // return_value_policy<return_by_value>(),
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/kernel/src/Exports/MaterialBuilder.cpp]: Any reason the comment is left around?

File: Framework/Kernel/inc/MantidKernel/Material.h [hunk 4]
```diff
@@ -50,21 +50,36 @@ File change history is stored at: <https://github.com/mantidproject/mantid>
 */
 class MANTID_KERNEL_DLL Material {
 public:
+  /// Structure to hold the information for a parsed chemical formula
+  struct FormulaUnit {
+    boost::shared_ptr<PhysicalConstants::Atom> atom;
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/Material.h]: Should we try and avoid the `shared_ptr` overhead with `unique_ptr` or do we want shared semantics?

File: Framework/Kernel/src/Material.cpp (hunk 2 above)
Comment (peterfpeterson) [inline Framework/Kernel/src/Material.cpp]: My current expectation is that the caller will cache the values. This may be incorrect.

Comment (martyngigg) [inline Framework/Kernel/src/Material.cpp]: They don't at the moment but they also pass the wavelength through, which is not yet used so if we started using that then we couldn't really cache anyway.

**Comment by martyngigg** [comment]
@peterfpeterson There are a couple of doyxgen warnings to deal with.

**Comment by martyngigg** [comment]
The code changes are great. `SetSampleMaterial` is now greatly simplified and the main workhorse code is now out of the algorithm and more accessible by the test of the code base.

I tried the script given and it produces sensible numbers. There are also errors thrown when input to the various builder methods is not what is expected.

Once the doxygen warnings are fixed I'm happy to pass it for merging.

**Comment by martyngigg** [comment]
Good to go. :shipit:

File: docs/source/release/v3.8.0/framework.rst [hunk 5]
```diff
@@ -76,6 +76,9 @@ Improved
 Python
 ------
 
+- :py:obj:`mantid.kernel.MaterialBuilder` has been exposed to python
+  and :py:obj:`mantid.kernel.Material` has been modified to expose the
```
Comment (FedeMPouzols) [inline docs/source/release/v3.8.0/framework.rst]: Nice cross-referencing in the release notes.

**Pull request merged**